### DISCUSSION TITLE
New Button to Convert $formulas$ in Note

### DISCUSSION THREAD

**Discussion by lhx332251335** [TriliumNext/Trilium#4792] [Show and tell]
### Introduction
This script is designed to convert math formulas wrapped in single dollar signs (`$`) and double dollar signs (`$$`) into HTML format within Trilium Notes. A new button is added to the center pane, which you can click to perform the conversion. Formulas wrapped in single dollar signs are treated as inline math, while those wrapped in double dollar signs are treated as block math. Please note that using this method for formula conversion cannot be undone using the Ctrl+Z shortcut.
![image](https://github.com/zadam/trilium/assets/35807793/5343f25f-78cf-4b93-9f17-673432e7f24e)
![image](https://github.com/zadam/trilium/assets/35807793/66323354-e532-4268-989c-c69522c7e69c)

### Usage Instructions
1. **Create a Script Note**:
   - Create a new note and change its type to **JS Frontend**.
   
2. **Copy the Script**:
   - Copy the following script into the newly created note:
```
class ReplaceFormulasButton extends api.NoteContextAwareWidget {

    get parentWidget() {
        return 'center-pane';
    }

    doRender() {
        this.$widget = $(`<style type="text/css">
            .replace-formulas-icon.ribbon-tab-title-icon.bx:before {
                content: "\\1D4C1"; 
            }
        </style>`);
        return this.$widget;
    }

    async refreshWithNote() {
        $(document).ready(() => {
            if (!$("div.component.note-split:not(.hidden-ext) div.ribbon-tab-title").hasClass('replace-formulas-button')) {
                $("div.component.note-split:not(.hidden-ext) .ribbon-tab-title:not(.backToHis)").last().after(`
                    <div class="replace-formulas-button ribbon-tab-spacer"></div>
                    <div class="replace-formulas-button ribbon-tab-title">
                        <span class="replace-formulas-icon ribbon-tab-title-icon bx" title="Replace Formulas"></span>
                    </div>
                `);
            }
            $('div.component.note-split:not(.hidden-ext) div.replace-formulas-button').off("click", replaceFormulas);
            $('div.component.note-split:not(.hidden-ext) div.replace-formulas-button').on("click", replaceFormulas);
        });
    }
}

var replaceFormulas = function () {
    api.getActiveContextTextEditor().then(editor => {
        const data = editor.getData();

        const regex = /\$\$(.*?)\$\$|\$(.*?)\$/gs;
        let replacedData = data;
        let matchFound = false;

        replacedData = replacedData.replace(regex, (match, p1, p2) => {
            matchFound = true;
            if (p1) {
                return `<span class="math-tex">\\[${p1.replace(/<br\s*\/?>/gi, '')}\\]</span>`;
            } else if (p2) {
                return `<span class="math-tex">\\(${p2.replace(/<br\s*\/?>/gi, '')}\\)</span>`;
            }
        });

        if (matchFound) {
            editor.setData(replacedData);
            api.showMessage('Formulas have been successfully replaced.');
        } else {
            api.showMessage('No formulas found to replace.');
        }
    }).catch(error => {
        console.error('Failed to get CKEditor instance:', error);
        api.showMessage('An error occurred while trying to replace formulas.');
    });
};

module.exports = new ReplaceFormulasButton();

```
3. **Add Widget Attribute**:
   - In the **Owned Attributes** section of the note, add `#widget`.

4. **Refresh Trilium**:
   - Press `Ctrl+Shift+R` to refresh Trilium Notes to apply the script.

**Comment by Nriver** [comment]
added to https://github.com/Nriver/awesome-trilium

**Comment by CsatiZoltan** [comment]
@lhx332251335 I don't see "JS Frontend" note type. Where do you set it?

**Reply by lhx332251335** [reply]
![image](https://github.com/user-attachments/assets/080805fb-37af-40af-973d-551a6dd09fbc)
If you don't have the option, click the trilium icon on the top-left, then "Options" - "Code Notes", and check JS frontend.
![image](https://github.com/user-attachments/assets/5e842e3d-a152-4b50-ae74-13aa6cdf2ed5)

**Reply by CsatiZoltan** [reply]
Thanks, it works after having activated in the settings.

It seems to have the same effect as using `Ctrl + m`.

**Comment by AlexCuan** [comment]
Hey, great work! 

In some notes it works for me, but in other it gives this error 
![image](https://github.com/user-attachments/assets/40eddf59-43a6-4e37-9cde-8261a7dd27dd)

And in others, even though there are several formulas, it says there are no formulas 
![image](https://github.com/user-attachments/assets/9b22b6c1-240e-4953-9d92-14cc00636d09)

I upload a sample file where it gives an error
[Path and surface integrals.md](https://github.com/user-attachments/files/18704267/Path.and.surface.integrals.md)

**Reply by AlexCuan** [reply]
Debugging by hand found that only happen in imported files, not in "handcopied" markdown

**Reply by lhx332251335** [reply]
Thank you for your feedback and for pointing out the issue. I tried to replicate the error you described, but unfortunately, I couldn't reproduce it on my end. Here's what I attempted:  

- **Copied Markdown text**, then used the *Markdown import from clipboard* button to import it into a note, followed by converting the formulas.  
- **Imported a note** by selecting an `.md` file directly, then converting the formulas.  

Both methods worked without any issues for me. I'm currently using **TriliumNext 0.91.6**.  

If there are any specific scenarios where the error occurs, or if you can provide additional steps to consistently reproduce the issue, please let me know, and I'll investigate further.

**Reply by AlexCuan** [reply]
Thanks you for the quick answer :) The problem doesn't arises when importing that way (sorry, I should have been more clear about that). The problem is when you import it through this way.

EDIT: I'm using the same version, tried at 2 browsers


![image](https://github.com/user-attachments/assets/594e8e13-9c78-42e0-806d-5b2fdfcf7529)

**Reply by lhx332251335** [reply]
I previously tested on the Windows client and just now tried the workflow using Chrome: importing the Markdown file in the exact way you described and then converting the formulas. However, I still couldn’t reproduce the issue on my side.

![2025-02-07-20363202502072059151722](https://github.com/user-attachments/assets/c71ce703-80ba-4193-9805-390964cdf38e)

**Reply by AlexCuan** [reply]
I'm sorry to hear that. It seems like a thing from my instance? I don't know. I've changed from browser, restarted the container, change of OS, and nothing.
[video (1).webm](https://github.com/user-attachments/assets/d7597190-084c-44f9-8382-74964573bc45)

I even tried on Firefox and it gives me this:

![image](https://github.com/user-attachments/assets/f783b494-8644-4866-9473-de0137567919)

And that link goes to:

![image](https://github.com/user-attachments/assets/b00520e4-a325-41c5-bed0-bb23501646f9)

**Comment by lhx332251335** [comment]
I’m really sorry to hear about the persistent issue. Unfortunately, as I
still cannot reproduce the error on my end, it’s difficult to pinpoint the
cause or provide specific script modifications. Based on the error code in
your screenshots, it seems the issue might stem from Trilium’s frontend API
not behaving as expected in your instance.
Here are a few suggestions that might help resolve or narrow down the
problem:
- It looks like you’re accessing Trilium through a reverse proxy. Could you
try connecting directly via the server’s IP? If it works without issues
this way, the problem might be related to your reverse proxy configuration.
Adjusting the settings there could resolve it.
- Backup your data files, then create a fresh container with a clean
Trilium instance to test the script. This will help determine if the issue
is specific to your current setup or environment.

**Reply by AlexCuan** [reply]
Tried, but the same happens. Even tried with a fresh chrome installing (until now I was using chromium) and nothing

**Comment by X-Bruce-Y** [comment]
Hi I try to convert `$LR_k = LR_{base} \times (\text{Decay Rate})^{\text{layer\_index}}$` to formula by your widget, it ends like:
<img width="420" height="59" alt="image" src="https://github.com/user-attachments/assets/e381ca5a-e1d1-4198-bc1e-46ab885e8c7a" />
Then I double click the formula, the popup suggests it was parsed as `\(LR_k = LR_{base} \times (`, which is wrong:
<img width="694" height="172" alt="image" src="https://github.com/user-attachments/assets/a817511d-8aae-403b-820e-ce72eec09fda" />
Then if I copy the original markdown formula into the popup and update it, it becomes the corrrect form:
<img width="865" height="174" alt="image" src="https://github.com/user-attachments/assets/d422ba17-bc9d-4bd1-b782-3406d1bd2660" />
Is there any change I should make to fix it? I'm using the exact code you provided on trilium 0.63.7 (95d46b7ce7e0036eeb72b43e1815cc5c00dc214c).

Thanks!

**Reply by eliandoran** [reply]
@X-Bruce-Y , the latest versions of Trilium already support direct Markdown import including for Math equations (you can use the "Import Markdown" button).

Is there a reason for using 0.63.7 (a version from May 2024)?

**Reply by lhx332251335** [reply]
Sorry, I’m not able to reproduce your error on Trilium v0.97.1. 

video: https://github.com/user-attachments/assets/11222863-3de8-4085-a538-0043fd96079d

As @eliandoran mentioned, I actually don't use this plugin very often myself.

**Reply by X-Bruce-Y** [reply]
> Sorry, I’m not able to reproduce your error on Trilium v0.97.1.
> 
> video: https://github.com/user-attachments/assets/11222863-3de8-4085-a538-0043fd96079d
> 
> As @eliandoran mentioned, I actually don't use this plugin very often myself.

Thanks a lot!

**Reply by X-Bruce-Y** [reply]
> @X-Bruce-Y , the latest versions of Trilium already support direct Markdown import including for Math equations (you can use the "Import Markdown" button).
> 
> Is there a reason for using 0.63.7 (a version from May 2024)?

Thank you for your reply. I gave it a quick try once and found a few of my widges (e.g. a schedule reminder) fail to run on TriliumNext. I'm no JS expert and quite occupied in recent months, so very likely I'll upgrade when having time to systematically migrate.

**Comment by ovo-Tim** [comment]
Since the UI changed a lot, the old version has stopped working.
Here is my new version:
``` js
const BTN_CLASS = "replace-formulas-ck-btn";
function replaceFormulas() {
    api.getActiveContextTextEditor()
        .then(editor => {
            if (!editor?.getData || !editor?.setData) {
                api.showMessage("No active text editor found.");
                return;
            }
            const data = editor.getData();
            const regex = /\$\$([\s\S]*?)\$\$|\$([\s\S]*?)\$/g;
            let found = false;
            const replaced = data.replace(regex, (match, blockFormula, inlineFormula) => {
                found = true;
                if (blockFormula !== undefined) {
                    const cleaned = blockFormula.replace(/<br\s*\/?>/gi, "");
                    return `<span class="math-tex">\\[${cleaned}\\]</span>`;
                }
                const cleaned = inlineFormula.replace(/<br\s*\/?>/gi, "");
                return `<span class="math-tex">\\(${cleaned}\\)</span>`;
            });
            if (!found) {
                api.showMessage("No formulas found to replace.");
                return;
            }
            editor.setData(replaced);
            api.showMessage("Formulas replaced.");
        })
        .catch(err => {
            console.error("replaceFormulas error:", err);
            api.showMessage("Failed to replace formulas.");
        });
}
class ReplaceFormulasWidget extends api.NoteContextAwareWidget {
    get parentWidget() {
        return "center-pane";
    }
    doRender() {
        this.$widget = $("<div></div>");
        return this.$widget;
    }
    removeButtons() {
        $(`button.${BTN_CLASS}`).off("click.replaceFormulas").remove();
    }
    getToolbarItems() {
        // Fixed/new layout toolbar target.
        let $items = $(".classic-toolbar-widget:not(.hidden-ext) .ck-toolbar__items").first();
        // Fallback for cases where hidden-ext isn't used.
        if (!$items.length) {
            $items = $(".classic-toolbar-widget .ck-toolbar__items").first();
        }
        return $items;
    }
    ensureButton() {
        const $items = this.getToolbarItems();
        if (!$items.length) return false;
        if ($items.find(`button.${BTN_CLASS}`).length) return true;
        const $btn = $(`
            <button
                class="ck ck-button ck-off ${BTN_CLASS}"
                type="button"
                tabindex="-1"
                aria-pressed="false"
                title="Replace formulas"
                data-cke-tooltip-text="Replace formulas"
                data-cke-tooltip-position="s"
            >
                <svg
                    class="ck ck-icon ck-reset_all-excluded ck-icon_inherit-color ck-button__icon"
                    viewBox="0 0 20 20"
                    aria-hidden="true"
                >
                    <path fill="currentColor" d="M4 4h12v2H7.2l5.8 6-5.8 6H16v2H4v-1.2L10.2 12 4 5.2V4z"></path>
                </svg>
                <span class="ck ck-button__label">Replace formulas</span>
            </button>
        `);
        $btn.on("click.replaceFormulas", e => {
            e.preventDefault();
            e.stopPropagation();
            replaceFormulas();
        });
        $items.append($btn);
        return true;
    }
    startWatchingToolbar() {
        if (this._observer) return;
        this._observer = new MutationObserver(() => {
            this.ensureButton();
        });
        this._observer.observe(document.body, {
            childList: true,
            subtree: true
        });
    }
    stopWatchingToolbar() {
        if (this._observer) {
            this._observer.disconnect();
            this._observer = null;
        }
    }
    async refreshWithNote(note) {
        // Keep button only on text notes.
        if (!note || note.type !== "text") {
            this.removeButtons();
            return;
        }
        this.startWatchingToolbar();
        this.ensureButton();
    }
    async entitiesReloadedEvent() {
        this.ensureButton();
    }
    async destroy() {
        this.stopWatchingToolbar();
        this.removeButtons();
        if (super.destroy) {
            await super.destroy();
        }
    }
}
module.exports = new ReplaceFormulasWidget();
```
<img width="292" height="126" alt="image" src="https://github.com/user-attachments/assets/060becf8-5b01-4393-a26f-202e9fe3c753" />
